### PR TITLE
SONAR-6195 Implement field selection in api/issues/search

### PR DESCRIPTION
**Pull request by jblievremont**

### CODE REVIEW

File: server/sonar-web/src/main/js/components/source-viewer/main.js [hunk 1]
```diff
@@ -274,7 +274,8 @@ define([
               options = {
                 data: {
                   componentUuids: this.model.id,
-                  extra_fields: 'actions,transitions,actionPlanName',
+                  f: 'component,componentId,project,subProject,rule,status,resolution,author,reporter,assignee,debt,line,message,' +
```
Comment (SonarTech) [inline server/sonar-web/src/main/js/components/source-viewer/main.js]: MINOR The line contains 132 characters which is greater than 120 authorized. [rule](http://nemo.sonarqube.org/coding_rules#rule_key=javascript%3ALineLength)

File: server/sonar-web/src/main/js/components/source-viewer/main.js [hunk 2]
```diff
@@ -274,7 +274,8 @@ define([
...
+                    'severity,actionPlan,creationDate,updateDate,closeDate,tags,comments,attr,actions,transitions,actionPlanName',
```
Comment (SonarTech) [inline server/sonar-web/src/main/js/components/source-viewer/main.js]: MINOR The line contains 130 characters which is greater than 120 authorized. [rule](http://nemo.sonarqube.org/coding_rules#rule_key=javascript%3ALineLength)

File: server/sonar-server/src/main/java/org/sonar/server/ws/JsonWriterUtils.java [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+package org.sonar.server.ws;
+
+import java.util.Collection;
+import javax.annotation.Nullable;
+import org.sonar.api.utils.text.JsonWriter;
+
+public class JsonWriterUtils {
```
Comment (SonarTech) [inline server/sonar-server/src/main/java/org/sonar/server/ws/JsonWriterUtils.java]: MAJOR Add a private constructor to hide the implicit public one. [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS1118)

File: server/sonar-web/src/main/js/apps/issues/controller.js [hunk 4]
```diff
@@ -5,7 +5,8 @@ define([
 ], function (Controller, ComponentViewer, HomeView) {
 
   var $ = jQuery,
-      EXTRA_FIELDS = 'actions,transitions,actionPlanName',
+      FIELDS = 'component,componentId,project,subProject,rule,status,resolution,author,reporter,assignee,debt,line,message,severity,' +
```
Comment (SonarTech) [inline server/sonar-web/src/main/js/apps/issues/controller.js]: MINOR The line contains 135 characters which is greater than 120 authorized. [rule](http://nemo.sonarqube.org/coding_rules#rule_key=javascript%3ALineLength)

File: server/sonar-server/src/main/java/org/sonar/server/issue/ws/IssueJsonWriter.java [hunk 5]
```diff
@@ -65,12 +99,22 @@ public IssueJsonWriter(I18n i18n, Durations durations, UserSession userSession,
   }
 
   public void write(JsonWriter json, Issue issue, Map<String, User> usersByLogin, Map<String, ComponentDto> componentsByUuid,
```
Comment (SonarTech) [inline server/sonar-server/src/main/java/org/sonar/server/issue/ws/IssueJsonWriter.java]: MAJOR The Cyclomatic Complexity of this method "write" is 13 which is greater than 10 authorized. [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AMethodCyclomaticComplexity)

File: server/sonar-server/src/main/java/org/sonar/server/issue/ws/IssueJsonWriter.java [hunk 6]
```diff
@@ -42,15 +45,46 @@
...
-    ACTIONS_EXTRA_FIELD, TRANSITIONS_EXTRA_FIELD, ACTION_PLAN_NAME_EXTRA_FIELD);
+    FIELD_ACTIONS, FIELD_TRANSITIONS, FIELD_ACTION_PLAN_NAME);
+
+  public static final Set<String> SELECTABLE_FIELDS = ImmutableSet.of(FIELD_COMPONENT, FIELD_PROJECT, FIELD_SUB_PROJECT, FIELD_RULE, FIELD_STATUS, FIELD_RESOLUTION, FIELD_AUTHOR,
+    FIELD_REPORTER, FIELD_ASSIGNEE, FIELD_DEBT, FIELD_LINE, FIELD_MESSAGE, FIELD_SEVERITY, FIELD_ACTION_PLAN, FIELD_CREATION_DATE, FIELD_UPDATE_DATE, FIELD_CLOSE_DATE,
+    FIELD_COMPONENT_ID, FIELD_TAGS, FIELD_COMMENTS, FIELD_ATTRIBUTES, FIELD_ACTIONS, FIELD_TRANSITIONS, FIELD_ACTION_PLAN_NAME);
+
+  private static final List<String> SELECTABLE_MINUS_EXTRAS = ImmutableList.copyOf(Sets.difference(SELECTABLE_FIELDS, EXTRA_FIELDS));
 
   private final I18n i18n;
```
Comment (SonarTech) [inline server/sonar-server/src/main/java/org/sonar/server/issue/ws/IssueJsonWriter.java]: MAJOR Remove this unused "i18n" private field. [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS1068)

**Comment by SonarTech** [comment]
SonarQube analysis reported 10 issues:
* MAJOR 7 major
* MINOR 3 minor

Watch the comments in this conversation to review them.
Note: the following issues could not be reported as comments because they are located on lines that are not displayed in this pull request:
* MAJOR Add a 'package-info.java' file to document the 'src/main/java/org/sonar/server/issue' package (org.codehaus.sonar:sonar-server:src/main/java/org/sonar/server/issue) [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS1228)
* MAJOR Add a 'package-info.java' file to document the 'src/main/java/org/sonar/server/issue/ws' package (org.codehaus.sonar:sonar-server:src/main/java/org/sonar/server/issue/ws) [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS1228)
* MAJOR Add a 'package-info.java' file to document the 'src/main/java/org/sonar/server/user/ws' package (org.codehaus.sonar:sonar-server:src/main/java/org/sonar/server/user/ws) [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS1228)
* MAJOR Add a 'package-info.java' file to document the 'src/main/java/org/sonar/server/ws' package (org.codehaus.sonar:sonar-server:src/main/java/org/sonar/server/ws) [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS1228)

**Comment by teryk** [comment]
please fix the analysis issues

File: server/sonar-server/src/main/java/org/sonar/server/issue/ws/IssueJsonWriter.java [hunk 7]
```diff
@@ -80,32 +124,37 @@ public void write(JsonWriter json, Issue issue, Map<String, User> usersByLogin,
...
-      .prop("creationDate", isoDate(issue.creationDate()))
-      .prop("updateDate", isoDate(updateDate))
-      .prop("closeDate", isoDate(issue.closeDate()));
-
-    writeTags(issue, json);
-    writeIssueComments(commentsByIssues.get(issue.key()), usersByLogin, json);
-    writeIssueAttributes(issue, json);
-    writeIssueExtraFields(issue, usersByLogin, actionPlanByKeys, extraFields, json);
+    json.prop(FIELD_KEY, issue.key());
+    JsonWriterUtils.writeIfNeeded(json, file != null ? file.getKey() : null, FIELD_COMPONENT, fields);
```
Comment (teryk) [inline server/sonar-server/src/main/java/org/sonar/server/issue/ws/IssueJsonWriter.java]: shouldn't it be in JsonWriter directly ?

File: server/sonar-server/src/main/java/org/sonar/server/ws/JsonWriterUtils.java [hunk 8]
```diff
@@ -0,0 +1,55 @@
...
+    }
+  }
+
+  public static void writeIfNeeded(JsonWriter json, @Nullable Long value, String field, Collection<String> fields) {
+    if (isFieldWanted(field, fields)) {
+      json.prop(field, value);
+    }
+  }
+
+  public static boolean isFieldWanted(String field, @Nullable Collection<String> fields) {
```
Comment (teryk) [inline server/sonar-server/src/main/java/org/sonar/server/ws/JsonWriterUtils.java]: isFieldNeeded to be consistent in naming. I'm not convinced with the writeIfNeeded, but good enough

**Comment by teryk** [comment]
Apply my minor feedback and LGTM

File: server/sonar-server/src/main/java/org/sonar/server/issue/ws/IssueJsonWriter.java (hunk 7 above)
Comment (jblievremont) [inline server/sonar-server/src/main/java/org/sonar/server/issue/ws/IssueJsonWriter.java]: I don't think so, ``JsonWriter`` is only about - well, writing to JSON. Knowing if a field should be written or not seems like another concern to me.

Comment (teryk) [inline server/sonar-server/src/main/java/org/sonar/server/issue/ws/IssueJsonWriter.java]: ok

**Pull request merged**